architecture/2025-01-serverless-feature-spec.md

```diff
@@ -0,0 +1,279 @@
+**Assumption**: The Radius model can be extended to support serverless compute platforms without significant changes to the existing model. Serverless platforms have orchestrators that expose APIs that can be used to model the resources in Radius. We have validated these assumptions by building a prototype for Azure Container Instances.
+
+**Assumption**: As a result of the implementation, the refactoring done to the core Radius codebase to support Serverless container runtimes should also allow for easier extensibility of Radius to support other new platforms going forward. In other words, the work done for Serverless should be genericized within the Radius code and not resemble custom implementations for Kubernetes and select Serverless platforms.
+
+**Question**: How do we handle the case where the underlying serverless platform (especially opinionated ones like Azure Container Apps or AWS Fargate) does not provide the necessary APIs to support the Radius model? We will answer this question by building prototypes for additional serverless platforms and evaluating the feasibility of supporting them in Radius.
+
+## Current state
+<!-- If we already have some ongoing investment in this area, summarize the current state and point to any relevant documents. -->
+There has been a proof-of-concept for ACI support in Radius built on top of the ACI nGroups features. Here is a demo of the prototype that has been created so far: https://youtu.be/NMNZE22nSQI?si=Dq7Q5WVKgHularsO&t=1201
+
+## Detailed user experience
+ <!-- <List of steps the user goes through from the start to the end of the scenario to provide more detailed view of exactly what the user is able to do given the new capabilities>  -->
+<!-- Step 1
+Step 2
+… -->
+
+### Step 1: Define and deploy a Radius Environment for serverless
+
+The user defines a new Radius Environment for serverless compute by creating a new Environment definition file (e.g. `env.bicep`) and specifying the necessary settings for the serverless platform before deploying the environment using `rad deploy env.bicep`.
+
+```diff
+resource environment 'Applications.Core/environments@2023-10-01-preview' = {
+  name: 'myenv'
+  properties: {
+    compute: {
++      kind: 'aci'   // Required. The kind of container runtime to use, e.g. 'aci' for Azure Container Instances and 'ecs' for AWS Elastic Container Service
++      namespace: 'default' // This is currently required, but may be made optional if it doesn't apply to all compute platforms
++      resourceGroup: '/subscriptions/.../resourceGroups/myrg' // This was introduced as a part of the ACI prototype, but may be changed depending on the implementation, need to figure out if namespace and resourceGroup should be combined into a single more generic property
+    }
++    providers: { // This was introduced as a part of the ACI prototype, but may be changed depending on the implementation
```

Comment:
What's the advantage of adding a provider to the Bicep template vs having a user register a provider via `rad init`? The provider would still be tracked by Radius in the case of `rad init`

Reply:
I think there can be a case where the provider for `rad init` might have a different scope than the `providers.scope` supplied in the `env.bicep`. For example, the scope of the cloud provider hosting the compute for a given environment might be `scope: '/subscriptions/${subscriptionId}/resourceGroups/${resourceGroup}'`. The provider credentials saved from a `rad init --full` command might have scope beyond this specific resource group and thus the same credentials can be used to deploy many environments with different scopes within the same provider subscription or account.

I will make a note of this nuance.